dpe/src/context.rs

```diff
@@ -150,70 +150,175 @@ pub(crate) struct ChildToRootIter<'a> {
     idx: usize,
     contexts: &'a [Context],
     done: bool,
+    count: usize,
 }
 
 impl ChildToRootIter<'_> {
     /// Create a new iterator that will start at the leaf and go to the root node.
-    pub fn new(leaf_idx: usize, contexts: &[Context]) -> Result<ChildToRootIter, DpeErrorCode> {
-        Ok(ChildToRootIter {
+    pub fn new(leaf_idx: usize, contexts: &[Context]) -> ChildToRootIter {
+        ChildToRootIter {
             idx: leaf_idx,
             contexts,
             done: false,
-        })
+            count: 0,
+        }
     }
 }
 
 impl<'a> Iterator for ChildToRootIter<'a> {
-    type Item = &'a Context;
+    type Item = Result<&'a Context, DpeErrorCode>;
 
-    fn next(&mut self) -> Option<&'a Context> {
+    fn next(&mut self) -> Option<Result<&'a Context, DpeErrorCode>> {
         if self.done {
             return None;
         }
+        if self.count >= MAX_HANDLES {
+            self.done = true;
+            return Some(Err(DpeErrorCode::MaxTcis));
+        }
 
         let context = &self.contexts[self.idx];
+
+        // Check if context is valid.
+        const MAX_IDX: u8 = (MAX_HANDLES - 1) as u8;
+        let valid_parent_idx = matches!(context.parent_idx, 0..=MAX_IDX | Context::ROOT_INDEX);
```

Comment:
Sorry, rust novice question. I can't quite figure out what this means. Is it:

`if parent_idx is in the range (0, MAX_IDX] or parent_idx is the ROOT_INDEX`?

Reply:
Really close, it is inclusive for the left part of the range as well. i.e.

```
if parent_idx is in the range [0, MAX_IDX] or parent_idx is the ROOT_INDEX
```

Reply:
Yes (modulo that 0 is also inclusive). Essentially, `matches!(expr, pattern)` is a sugar for:

```rust
match expr {
    pattern => true,
    _ => false,
}
```

Reply:
To expand this macro as @ia0 mentioned, this would read:

```Rust
let valid_parent_idx = match context.parent_idx {
    0..=MAX_IDX | Context::ROOT_INDEX => true,
    _ => false,
};
```

For the pipe this would expand it a little further to look something like this:

```Rust
let valid_parent_idx = match context.parent_idx {
    0..=MAX_IDX => true,
    Context::ROOT_INDEX => true,
    _ => false,
};
```

Reply:
Thanks!